From d4bb0274f52e2d013c20bd42efa2ec45ae7fcb34 Mon Sep 17 00:00:00 2001
From: Santtu Lakkala <santtu.lakkala@unikie.com>
Date: Tue, 28 Oct 2025 07:37:22 +0400
Subject: [PATCH] localed: use LOCALE_ARCHIVE

Allow specifying locale archive location via LOCALE_ARCHIVE environment variable.
This is useful for NixOS and other immutable distributions.

Signed-off-by: Santtu Lakkala <santtu.lakkala@unikie.com>
Patch applied from Ghaf
Signed-off-by: Brian McGillion <bmg.avoin@gmail.com>
---
 src/basic/locale-util.c          | 6 +++++-
 units/systemd-localed.service.in | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/basic/locale-util.c b/src/basic/locale-util.c
index 455b54bc61..ff0294226a 100644
--- a/src/basic/locale-util.c
+++ b/src/basic/locale-util.c
@@ -92,7 +92,11 @@ static int add_locales_from_archive(Set *locales) {

         assert(locales);

-        _cleanup_close_ int fd = open("/usr/lib/locale/locale-archive", O_RDONLY|O_NOCTTY|O_CLOEXEC);
+        const char *archive = secure_getenv("LOCALE_ARCHIVE");
+        if (!archive)
+                archive = "/usr/lib/locale/locale-archive";
+
+        _cleanup_close_ int fd = open(archive, O_RDONLY|O_NOCTTY|O_CLOEXEC);
         if (fd < 0)
                 return errno == ENOENT ? 0 : -errno;

diff --git a/units/systemd-localed.service.in b/units/systemd-localed.service.in
index 4de89aa8dd..1e72b18edc 100644
--- a/units/systemd-localed.service.in
+++ b/units/systemd-localed.service.in
@@ -45,4 +45,5 @@ RestrictSUIDSGID=yes
 SystemCallArchitectures=native
 SystemCallErrorNumber=EPERM
 SystemCallFilter=@system-service
+Environment=LOCALE_ARCHIVE=/run/current-system/sw/lib/locale/locale-archive
 {{SERVICE_WATCHDOG}}
--
2.51.0
